### ISSUE TITLE
[Feature Request] Add a Helper for Pagination Response Support

### ISSUE THREAD

**Issue by muco-rolle** [adocasts/package-dto#1]
Hello,

In the generated DTO for a resource, there's a fromArray helper function. I believe it would be beneficial for it to support converting paginated responses since they include different fields (meta and data). Alternatively, we could add another helper function like fromPaginator.

I'm not entirely sure how to implement this, but I'm open to submitting a pull request with some guidance.

I have a working, albeit hacky, solution in my project. Please let me know how it can be improved.

Here is how the JobDto looks, including the fromPaginator helper function:

```ts
import Job from '#models/job';
import { SimplePaginator } from '@adonisjs/lucid/database';
import { ModelPaginatorContract } from '@adonisjs/lucid/types/model';

type SimplePaginatorMeta = {
  total: number;
  perPage: number;
  currentPage: number;
  lastPage: number;
  firstPage: number;
  firstPageUrl: string | null;
  lastPageUrl: string | null;
  nextPageUrl: string;
  previousPageUrl: string | null;
};

export default class JobDto {
  declare id: number;
  declare title: string;
  declare salary: string;
  declare createdAt: string;

  constructor(job?: Job) {
    if (!job) return;

    this.id = job!.id;
    this.title = job!.title;
    this.salary = job!.salary;
    this.createdAt = job!.createdAt.toISO()!;
  }

  static fromArray(jobs: Job[]) {
    if (!jobs) return [];

    return jobs.map((job) => new JobDto(job));
  }

  static fromPaginator(jobs: ModelPaginatorContract<Job>) {
    if (!(jobs instanceof SimplePaginator)) {
      throw new Error('jobs is not an instance of SimplePaginator');
    }

    const data = jobs.toJSON().data as Job[];
    const meta = jobs.toJSON().meta as SimplePaginatorMeta;

    return {
      data: data.length ? data.map((job) => new JobDto(job)) : [],
      meta,
    };
  }
}
```

**Comment by muco-rolle** [comment]
Hello @tomgobich. 

Do you have any input on the above?

**Comment by tomgobich** [comment]
Hey Muco! Terribly sorry, this slipped past my radar!

Offering a solution for pagination is a good idea! 

I'm kind of going back and forth at the moment on two different approaches, and it's getting late so I'll dump my current thinking here and return back tomorrow after work. Feel free to chime in if you have any thoughts! :) 

**Thought 1:**
Perhaps we extract out a `BaseDto` that the then generated DTOs can extend. On the `BaseDto` is where we'd have these utility methods, which would also allow us to have a more centralized version of something like what you have above.

I'm not crazy about how the generics needed to do this looks though as the `BaseDto` would likely need to be:
```ts
class BaseDto<Model, Dto extends new (model: Model) => any> {
  constructor(protected dto: Dto) {}

  // other stuff
}
```
```ts
export default class DifficultyDto extends BaseDto<Difficulty, typeof DifficultyDto> {
  // ...declares

  constructor(difficulty?: Difficulty) {
    super(DifficultyDto)
  }
}
```

**Thought 2:**
After seeing the result of the `BaseDto` I started thinking that it might be cleaner to instead try to do a `SimplePaginatorDto` that accepts the model paginator.

Something that'd look like the below. Not sure how the generics of this would look though, I didn't manage to get that far tonight.
```ts
const paginated = await Job.query().paginate(1, 10)
const result = new SimplePaginatorDto(paginated)
```

I'm trying to think if there are other utilities that may result in thought 1 being more beneficial.

**Comment by tomgobich** [comment]
Circling back around to this today, I think I've got it figured out where we can offer both a `BaseDto` containing the utilities and a separate `SimplePaginatorDto` option.

I'll try to add what I've got to the package sometime this week! If all goes according to plan, we'll be able to do:

```ts
export default class DifficultyDto extends BaseDto {
  declare id: number
  declare organizationId: number
  declare name: string
  declare color: string
  declare order: number
  declare isDefault: boolean
  declare createdAt: string
  declare updatedAt: string
  declare organization: OrganizationDto | null

  constructor(difficulty?: Difficulty) {
    super()

    if (!difficulty) return
    this.id = difficulty.id
    this.organizationId = difficulty.organizationId
    this.name = difficulty.name
    this.color = difficulty.color
    this.order = difficulty.order
    this.isDefault = difficulty.isDefault
    this.createdAt = difficulty.createdAt.toISO()!
    this.updatedAt = difficulty.updatedAt.toISO()!
    this.organization = difficulty.organization && new OrganizationDto(difficulty.organization)
  }
}

const difficulties = await organization.getDifficulties()
const dtos = DifficultyDto.fromArray(difficulties)

const paginator = await Difficulty.query().paginate(1, 10)
const pages = DifficultyDto.fromPaginator(paginator, { start: 1, end: 3 })
```

I may tweak this a little further yet, but working off what you have above, below is where my current thinking is:
```ts
interface SimplePaginatorMetaContract {
  total: number
  perPage: number
  currentPage: number
  lastPage: number
  firstPage: number
  firstPageUrl: string
  lastPageUrl: string
  nextPageUrl: string | null
  previousPageUrl: string | null
  pagesInRange?: {
    url: string
    page: number
    isActive: boolean
  }[]
}

interface SimplePaginatorMetaRangeContract {
  start: number
  end: number
}

class SimplePaginatorDto<Model extends LucidRow, Dto extends new (model: Model) => any> {
  declare data: Dto[]
  declare meta: SimplePaginatorMetaContract

  constructor(
    paginator: ModelPaginatorContract<Model>,
    dto: Dto,
    range?: SimplePaginatorMetaRangeContract
  ) {
    this.data = paginator.all().map((row) => new dto(row))
    this.meta = {
      total: paginator.total,
      perPage: paginator.perPage,
      currentPage: paginator.currentPage,
      lastPage: paginator.lastPage,
      firstPage: paginator.firstPage,
      firstPageUrl: paginator.getUrl(1),
      lastPageUrl: paginator.getUrl(paginator.lastPage),
      nextPageUrl: paginator.getNextPageUrl(),
      previousPageUrl: paginator.getPreviousPageUrl(),
    }

    if (range?.start || range?.end) {
      const start = range?.start || paginator.firstPage
      const end = range?.end || paginator.lastPage

      this.meta.pagesInRange = paginator.getUrlsForRange(start, end)
    }
  }
}

export type StaticThis<Model, Dto> = { new (model: Model): Dto }

class BaseDto {
  static fromArray<Model, Dto extends BaseDto>(this: StaticThis<Model, Dto>, models: Model[]) {
    if (!models) return []
    return models.map((model) => new this(model))
  }

  static fromPaginator<Model extends LucidRow, Dto extends BaseDto>(
    this: StaticThis<Model, Dto>,
    paginator: ModelPaginatorContract<Model>,
    range?: SimplePaginatorMetaRangeContract
  ) {
    return new SimplePaginatorDto(paginator, this, range)
  }
}
```

**Comment by muco-rolle** [comment]
This is excellent. The concept of BaseDto is great, as it opens up more opportunities for extensibility and building upon it.

I want to help, but I lack experience with Generics. I'll keep following along, so let me know if there's anything I can help.

Thank you so much.

**Comment by tomgobich** [comment]
Alrighty, this should be good to go! I published it last night, but I had a weird issue where the type inference worked locally with `npm link` but didn't once I published it and used it from NPM.

Think I got that fixed up, and the working version is published as `0.0.6`. When/if you get a chance to give it a go, let me know if you run into any issues!

There's now a `BaseDto` that contains the `fromArray` helper, as it isn't model specific and a `BaseModelDto` that adds the `fromPaginator`.

Here's an example DTO
```ts
import Difficulty from '#models/difficulty'
import OrganizationDto from '#dtos/organization'
import { BaseModelDto } from '@adocasts.com/dto/base'

export default class DifficultyDto extends BaseModelDto {
  declare id: number
  declare organizationId: number
  declare name: string
  declare color: string
  declare order: number
  declare isDefault: boolean
  declare createdAt: string
  declare updatedAt: string
  declare organization: OrganizationDto | null

  constructor(difficulty?: Difficulty) {
    super()

    if (!difficulty) return
    this.id = difficulty.id
    this.organizationId = difficulty.organizationId
    this.name = difficulty.name
    this.color = difficulty.color
    this.order = difficulty.order
    this.isDefault = difficulty.isDefault
    this.createdAt = difficulty.createdAt.toISO()!
    this.updatedAt = difficulty.updatedAt.toISO()!
    this.organization = difficulty.organization && new OrganizationDto(difficulty.organization)
  }
}
```
Example controller method
```ts
import DifficultyDto from '#dtos/difficulty'
import Difficulty from '#models/difficulty'
//
async index({ inertia, organization }: HttpContext) {
  const difficulties = await organization.getDifficulties()

  const paginator = await Difficulty.query().paginate(1, 10)
  const paginated = DifficultyDto.fromPaginator(paginator)

  return inertia.render('difficulties/index', {
    difficulties: DifficultyDto.fromArray(difficulties),
    paginated,
  })
}
```
And, example Inertia props (Vue)
```ts
import DifficultyDto from '#dtos/difficulty'
import { SimplePaginatorDtoContract } from '@adocasts.com/dto/types'

const props = defineProps<{
  difficulties: DifficultyDto[]
  organization: OrganizationDto
  paginated: SimplePaginatorDtoContract<DifficultyDto>
}>()

props.paginated.data[0].name
props.paginated.meta.total
```

<img width="1106" alt="Screenshot 2024-07-25 at 7 56 52 AM" src="https://github.com/user-attachments/assets/c4e131b8-4ee0-4557-92e3-182085460b9d">
